tests/test_cli_compile.py

```diff
@@ -2735,9 +2737,383 @@ def test_all_extras_fail_with_extra(fake_dists, runner, make_module, fname, cont
+                    [
+                        (
+                            "small-fake-with-build-deps"
+                            " (pyproject.toml::build-system.backend::editable)"
+                        ),
+                        "small-fake-with-build-deps (pyproject.toml::build-system.backend::sdist)",
+                        "small-fake-with-build-deps (pyproject.toml::build-system.backend::wheel)",
+                    ],
+                ),
+                "fake-dynamic-build-dep-for-sdist": Dependency(
+                    "0.3",
+                    [
+                        "small-fake-with-build-deps (pyproject.toml::build-system.backend::sdist)",
+                    ],
+                ),
+                "fake-dynamic-build-dep-for-wheel": Dependency(
+                    "0.4",
+                    [
+                        "small-fake-with-build-deps (pyproject.toml::build-system.backend::wheel)",
+                    ],
+                ),
+                "fake-dynamic-build-dep-for-editable": Dependency(
+                    "0.5",
+                    [
+                        (
+                            "small-fake-with-build-deps"
+                            " (pyproject.toml::build-system.backend::editable)"
+                        ),
+                    ],
+                ),
+                "small-fake-a": Dependency(
+                    "0.1",
+                    [
+                        "small-fake-with-build-deps (setup.py)",
+                    ],
+                ),
+                "small-fake-b": Dependency(
+                    "0.2",
+                    [
+                        "small-fake-with-build-deps (setup.py)",
+                    ],
+                ),
+                "wheel": Dependency(
+                    "*",
+                    [
+                        "small-fake-with-build-deps (pyproject.toml::build-system.backend::wheel)",
+                    ],
+                ),
+            },
+        ),
+    ),
+)
+def test_build_distribution(
+    fake_dists_with_build_deps,
+    runner,
+    tmp_path,
+    monkeypatch,
+    distributions,
+    other_options,
+    expected_deps,
+    make_package,
+    make_wheel,
+):
+    """
+    Test that when one or more ``--build-deps-for`` are given the expected packages are
+    included.
+    """
+    # When used as argument to the runner it is not passed to pip
+    monkeypatch.setenv("PIP_FIND_LINKS", fake_dists_with_build_deps)
+    src_pkg_path = os.path.join(PACKAGES_PATH, "small_fake_with_build_deps")
+    base_cmd = ["-n", "--allow-unsafe"]
+
+    with runner.isolated_filesystem(tmp_path) as tmp_pkg_path:
+        shutil.copytree(src_pkg_path, tmp_pkg_path, dirs_exist_ok=True)
+        out = runner.invoke(
+            cli,
+            base_cmd + [f"--build-deps-for={d}" for d in distributions] + other_options,
+        )
+    assert out.exit_code == 0
+    assert _parse_deps(out.stderr, expected_deps) == expected_deps
```

Comment:
I would suggest asserting the expected raw output rather than the computed `_parse_deps(...)`. I appreciate the concise and DRY-oriented approach of the current test, but it might be a bit challenging to understand without a deep understanding of the `_parse_deps` function.

Please take a look at this [example of an overengineered test](https://github.com/jazzband/pip-tools/blob/81ecd9bbc43a3fc1a82ac4de42e2aa92d29ce0cd/tests/test_resolver.py#L11-L206). It represents the kind of complexity I would avoid.

Reply:
At first it was parsing only the package versions and seemed like a neat way to get around the problem of setuptools not getting a reproducible version. But I agree it has grown a bit unwieldy and I will look into another way to make the tests reproducible.

Reply:
I think this has been resolved. Not sure why I did not think of pinning the package version before.